File: #0143.reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {

        vector<ListNode*> v;
        ListNode* ln = head;
        int i = 0;
        int j = 0;
        while (ln) {
            v.push_back(ln);
            j++;
            ln = ln->next;
        }
        j--;

        while (i < j) {
            v[i]->next = v[j];
            i++;
            if (i<j) {
                v[j]->next = v[i];
            }
            j--;
        }
        v[i]->next = NULL;
    }
};

File: #0143.reorder-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    let nodes = [];
    let node = head;
    while (node) {
        nodes.push(node);
        node = node.next;
    }

    nodes = nodes.slice(Math.ceil(nodes.length / 2));
    node = head;
    while (nodes.length) {
      const nextNode = nodes.pop();
      let tmp = node.next;
      node.next = nextNode;
      nextNode.next = tmp;
      node = tmp;
    }

    node.next = null;
    return head;
};
